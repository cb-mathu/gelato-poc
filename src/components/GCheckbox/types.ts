import { InputHTMLAttributes, ExtractPropTypes } from 'vue';
import { VueComponentProps } from '../../types';
import { CHECKBOX_PROPS } from './constants';

type Value = boolean | number | string | undefined;

export type CheckboxProps = ExtractPropTypes<typeof CHECKBOX_PROPS> & {
  ['true-value']?: Value;
  ['false-value']?: Value;
};
export type NativeButtonProps = Omit<InputHTMLAttributes, keyof CheckboxProps>;
export type MergedCheckboxProps = Partial<CheckboxProps & VueComponentProps & NativeButtonProps>;
