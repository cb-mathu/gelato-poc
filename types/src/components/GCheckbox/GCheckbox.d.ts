import { MergedCheckboxProps } from './types';
declare const _default: new () => {
    $props: MergedCheckboxProps;
};
export default _default;
