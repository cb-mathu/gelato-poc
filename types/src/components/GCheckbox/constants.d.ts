export declare const CHECKBOX_PROPS: {
    trueValue: {
        type: (StringConstructor | BooleanConstructor | NumberConstructor)[];
        default: boolean;
    };
    falseValue: {
        type: (StringConstructor | BooleanConstructor | NumberConstructor)[];
        default: boolean;
    };
    value: {
        type: (StringConstructor | NumberConstructor)[];
    };
    label: {
        type: StringConstructor;
    };
    description: {
        type: StringConstructor;
    };
    modelValue: {
        type: (StringConstructor | BooleanConstructor | NumberConstructor | ObjectConstructor | ArrayConstructor)[];
    };
    id: {
        type: StringConstructor;
        required: boolean;
    };
    class: {
        type: StringConstructor;
        required: boolean;
        default: string;
    };
    disabled: {
        type: BooleanConstructor;
    };
    autofocus: {
        type: BooleanConstructor;
        default: boolean;
    };
    autocomplete: {
        type: StringConstructor;
        default: string;
    };
    required: {
        type: BooleanConstructor;
        default: boolean;
    };
    readonly: {
        type: BooleanConstructor;
        default: boolean;
    };
    name: {
        type: StringConstructor;
        default: string;
    };
    placeholder: {
        type: StringConstructor;
        default: string;
    };
    modelModifiers: {
        default: () => {};
    };
    invalid: {
        type: BooleanConstructor;
    };
    errorMessage: {
        type: StringConstructor;
    };
};
