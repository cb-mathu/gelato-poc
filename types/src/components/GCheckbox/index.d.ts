import GCheckbox from './GCheckbox';
export { GCheckbox };
