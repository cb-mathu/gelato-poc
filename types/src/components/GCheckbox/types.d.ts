import { InputHTMLAttributes, ExtractPropTypes } from 'vue';
import { VueComponentProps } from '../../types';
import { CHECKBOX_PROPS } from './constants';
declare type Value = boolean | number | string | undefined;
export declare type CheckboxProps = ExtractPropTypes<typeof CHECKBOX_PROPS> & {
    ['true-value']?: Value;
    ['false-value']?: Value;
};
export declare type NativeButtonProps = Omit<InputHTMLAttributes, keyof CheckboxProps>;
export declare type MergedCheckboxProps = Partial<CheckboxProps & VueComponentProps & NativeButtonProps>;
export {};
