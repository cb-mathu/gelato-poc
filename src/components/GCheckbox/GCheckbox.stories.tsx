import { ref } from 'vue';
import GCheckbox from './GCheckbox';
import { MergedCheckboxProps } from './types';

export default {
  title: 'Component/Forms/GCheckbox',
  component: GCheckbox,
  argTypes: {
    name: {
      default: 'Name of the field',
    },
    label: {
      description: 'Label for the field',
      control: {
        type: 'text',
      },
    },
    disabled: {
      description: 'Disable the field',
      control: {
        type: 'boolean',
      },
    },
    autofocus: {
      description: 'Autofocus the field',
      default: false,
      control: {
        type: 'boolean',
      },
    },
    required: {
      description: 'Field is required',
      default: false,
      control: {
        type: 'boolean',
      },
    },
    readonly: {
      description: 'Field is readonly',
      default: false,
      control: {
        type: 'boolean',
      },
    },
    modelValue: {
      description:
        "v-model value: If it's a single checkbox, modelValue will be either trueValue or falseValue. If its group/multi checkbox, modelValue will be array of checked values",
      defaultValue: 'consent1',
      control: {
        type: 'text',
      },
    },
    trueValue: {
      description: 'Value to be emitted when the field is checked. Applicable only for single checkbox. Eg: "yes"',
      defaultValue: 'consent1',
      control: {
        type: 'text',
      },
    },
    falseValue: {
      description: 'Value to be emitted when the field is unchecked. Applicable only for single checkbox. Eg: "no"',
      defaultValue: 'consent2',
      control: {
        type: 'text',
      },
    },
    value: {
      description: 'Default value',
      control: {
        type: 'text',
      },
    },
    id: {
      description: 'Id of the field',
      control: {
        type: 'text',
      },
    },
    class: {
      description: 'Class of the field',
      defaulValue: '',
      control: {
        type: 'text',
      },
    },
    invalid: {
      description: 'Field is invalid',
      defaulValue: false,
      control: {
        type: 'boolean',
      },
    },
    errorMessage: {
      description: 'Error message for invalid field (Applicable only if invalid is set)',
      defaulValue: '',
      control: {
        type: 'text',
      },
    },
  },
};

const Template = (args: any) => ({
  components: { GCheckbox },
  setup() {
    const consentG1 = 'consent1';
    const consentG2 = 'consent2';
    const input = ref([consentG1]);
    const solo = ref(args.modelValue || consentG1);

    const attributes = {
      disabled: args.disabled,
      readonly: args.readonly,
      invalid: args.invalid,
      errorMessage: args.errorMessage,
    };

    return () => {
      const slots = {
        label: () => <span>{args.label}</span>,
      };
      return (
        <div style="display: grid; grid-gap: 0.5rem;">
          <b>Selected: {solo.value}</b>
          <div>
            <GCheckbox
              {...attributes}
              id="sep-checkbox-story"
              v-model={solo.value}
              trueValue={consentG1}
              falseValue={consentG2}
              v-slots={slots}
            />
          </div>
          <hr />
          <b>Group selected: {input.value.join(', ')}</b>

          <GCheckbox {...attributes} id="checkbox-story1" v-model={input.value} value={consentG1} label={args.groupLabel1} />
          <GCheckbox {...attributes} id="checkbox-story2" v-model={input.value} value={consentG2} label={args.groupLabel2} />
        </div>
      );
    };
  },
});

const slots = {
  label: () => <span>Ship to my billing address</span>,
};

export const Story: any = Template.bind({});
Story.args = {
  label: 'I would like to receive marketing emails from Acme Inc.',
  groupLabel1: 'I would like to receive marketing emails from Acme Inc.',
  groupLabel2: 'Send me information about new products and services when available.',
};

export const DefaultCheckbox = () => <GCheckbox />;

export const CheckboxWithLabel = (props?: MergedCheckboxProps) => <GCheckbox {...(props as any)} label="Ship to my billing address" />;

export const CheckboxWithLabelAsSlot = () => <GCheckbox v-slots={slots} />;

export const DisabledCheckbox = () => <GCheckbox label="Ship to my billing address" disabled />;

export const ReadonlyCheckbox = () => <GCheckbox label="Ship to my billing address" readonly />;

export const InvalidCheckbox = () => (
  <GCheckbox label="Ship to my billing address" invalid errorMessage="This is an error message, ooh poor you!" />
);

export const CheckboxGroup = (props?: any) => {
  return {
    setup() {
      const addresses = ref(props?.addresses || []);
      return () => (
        <>
          <div>Selected : [{addresses.value?.join(', ')}]</div>
          <GCheckbox id="checkbox-test1" v-model={addresses.value} value="billing" label="Ship to my billing address" />
          <GCheckbox id="checkbox-test2" v-model={addresses.value} value="shipping" label="Ship to my shipping address" />
        </>
      );
    },
  };
};
