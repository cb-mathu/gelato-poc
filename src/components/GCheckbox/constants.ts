import { BASE_FORM_PROPS } from '../../utils/props';

export const CHECKBOX_PROPS = {
  ...BASE_FORM_PROPS,
  trueValue: {
    type: [Boolean, Number, String],
    default: true,
  },
  falseValue: {
    type: [Boolean, Number, String],
    default: false,
  },
  value: {
    type: [String, Number],
  },
  label: {
    type: String,
  },
  description: {
    type: String,
  },
  modelValue: {
    type: [String, Number, Object, Array, Boolean],
  },
};
