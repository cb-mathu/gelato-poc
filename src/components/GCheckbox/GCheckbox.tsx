import { defineComponent, computed, ref, Ref, InputHTMLAttributes, watchEffect } from 'vue';
import { CHECKBOX_PROPS } from './constants';
import { useId, useVModel } from '../../composables';
import { MergedCheckboxProps } from './types';
import { useControlAria, useFormAria, useFormAriaIds } from '../../utils/accessibility';
import { GInputErrorMessage } from '../GInput';
import { isNotEmpty } from '../../utils/common';

const GCheckbox = defineComponent({
  name: 'GCheckbox',
  props: CHECKBOX_PROPS,

  setup(props, { slots, attrs }) {
    const checkbox: Ref<HTMLInputElement | null> = ref(null);
    const internalValue = useVModel(props, 'modelValue');
    const { labelId, errorId } = useFormAriaIds(props.id);

    return () => {
      const checked = computed(() => {
        if (props.modelValue instanceof Array) {
          return props.modelValue.includes(props.value);
        }
        return props.modelValue === props.trueValue;
      }).value;

      const indeterminate = computed(() => checkbox.value?.indeterminate).value;
      const label = props.label || slots.label?.();
      const propsWeControl = {
        id: props.id || useId('element'),
        ['true-value']: props.trueValue,
        ['false-value']: props.falseValue,
      };
      const attributes = { ...attrs, ...props, ...propsWeControl };
      const ariaAttributes = {
        ...useFormAria({
          labelledBy: labelId,
          invalid: props.invalid,
          errorMessage: errorId,
        }),
        ...useControlAria({
          checked,
          indeterminate,
        }),
      } as InputHTMLAttributes;
      const classes = {
        'g-checkbox': true,
        'g-checkbox--invalid': props.invalid,
        'g-checkbox--disabled': props.disabled,
        [`group-${props.id}`]: props.id,
      };

      return (
        <div class={classes}>
          <input
            ref={checkbox}
            checked={checked}
            type="checkbox"
            {...attributes}
            {...ariaAttributes}
            v-model={internalValue.value}
            class={`g-checkbox__element checkbox_${attributes.id}`}
          />
          {isNotEmpty(checkbox) && (
            <label class={`g-checkbox__label label_${attributes.id}`} for={attributes.id} id={labelId}>
              {label}
            </label>
          )}

          {props.invalid && props.errorMessage && (
            <GInputErrorMessage id={errorId} class="g-checkbox__error" errorMessage={props.errorMessage} />
          )}
        </div>
      );
    };
  },
});

export default GCheckbox as {
  new (): {
    $props: MergedCheckboxProps;
  };
};
