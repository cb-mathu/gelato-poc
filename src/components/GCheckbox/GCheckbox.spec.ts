import { render } from '@testing-library/vue';
import { screen } from '@testing-library/dom';
import userEvent from '@testing-library/user-event';
import * as stories from './GCheckbox.stories';
import { ref } from 'vue';

const {
  DefaultCheckbox,
  DisabledCheckbox,
  CheckboxWithLabel,
  CheckboxWithLabelAsSlot,
  InvalidCheckbox,
  ReadonlyCheckbox,
  CheckboxGroup,
} = stories;

const testLabel = 'Ship to my billing address';
const testErroMessage = 'This is an error message, ooh poor you!';

describe('GCheckbox : ', () => {
  test('renders with default args', () => {
    const { container } = render(DefaultCheckbox());
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders with label passed via prop', () => {
    render(CheckboxWithLabel());
    expect(screen.getByText(testLabel)).toBeVisible();
  });

  test('renders with label passed via slot', () => {
    render(CheckboxWithLabelAsSlot());
    expect(screen.getByText(testLabel)).toBeVisible();
  });

  test('renders in disabled state', () => {
    render(DisabledCheckbox());
    expect(screen.getByRole('checkbox')).toHaveAttribute('disabled');
  });

  test('renders in readonly state', () => {
    render(ReadonlyCheckbox());
    expect(screen.getByRole('checkbox')).toHaveAttribute('readonly');
  });

  test('renders in invalid state', () => {
    render(InvalidCheckbox());
    expect(screen.getByText(testErroMessage)).toBeVisible();
  });

  test('pressing tab will focus checkbox', () => {
    render(CheckboxWithLabel());
    userEvent.tab();
    expect(screen.getByRole('checkbox')).toHaveFocus();
  });

  test('focuses checkbox when label is clicked', () => {
    render(CheckboxWithLabel());
    const element = screen.getByText(testLabel);
    userEvent.click(element);
    expect(screen.getByRole('checkbox')).toHaveFocus();
  });

  test('does not update value in disabled state when label is clicked', () => {
    render(CheckboxWithLabel({ disabled: true }));
    userEvent.click(screen.getByText(testLabel));
    expect(screen.getByRole('checkbox')).not.toBeChecked();
  });

  test('updates the state on click', async () => {
    render(CheckboxWithLabel());
    await userEvent.click(screen.getByRole('checkbox'));
    expect(screen.getByRole('checkbox')).toBeChecked();
  });
});

describe('GCheckbox Group : ', () => {
  test('updates the value when binded to an array', async () => {
    const addresses = ref([] as string[]);
    const { container } = render(CheckboxGroup({ addresses }));
    await userEvent.click(screen.getByText(testLabel));
    expect(addresses.value.includes('billing')).toBe(true);
  });
});
