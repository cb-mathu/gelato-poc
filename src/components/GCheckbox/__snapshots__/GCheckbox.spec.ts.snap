// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GCheckbox :  renders with default args 1`] = `
<div
  class="g-checkbox"
>
  <input
    aria-checked="false"
    aria-errormessage=""
    aria-invalid="false"
    aria-labelledby="label_1"
    autocomplete="off"
    class="g-checkbox__element checkbox_element_4"
    false-value="false"
    falsevalue="false"
    id="element_4"
    invalid="false"
    modelmodifiers="[object Object]"
    modelvalue="false"
    name=""
    placeholder=""
    true-value="true"
    truevalue="true"
    type="checkbox"
  />
  <label
    class="g-checkbox__label label_element_4"
    for="element_4"
    id="label_1"
  >
    <!---->
  </label>
  <!---->
</div>
`;
